Support saving and loading ShardedTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62618
* __->__ #62242

1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

pritamdamania committed:
Support saving and loading ShardedTensor.

1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -472,6 +511,76 @@ def test_insufficient_sharding_dims(self):
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_state_dict(self):
```
bookmark

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -423,3 +470,71 @@ def remote_shards(self) -> Dict[int, List[rpc.RRef[Shard]]]:
+                '"init_process_group" before loading ShardedTensor')
+
+        self._local_shards, self._metadata, pg_state, self._sharding_spec = state
+        self._rpc_initialized = False
```
should we also recover this state? a sharded tensor have rpc initialized got saved, and reloading should also have rpc initialized?

pritamdamania87 replied:
It is automatically initialized below when we call `init_rpc`. I think we might have to rework this part anyways, since it seems like users would like this behavior to be explicit where we set up RRefs only if requested via `init_rrefs` parameter in the ShardedTensor constructor.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -235,7 +283,7 @@ def _init_rpc(self):
 
         # Barrier for all RPCs to finish on all ranks.
-        rpc.api._barrier(rpc_workers)
+        rpc.api._all_gather(None)
```
why not a barrier anymore?

pritamdamania87 replied:
As part of this PR, I'm removing the requirement for sub process groups where you need to do something like:

```
if self.rank >= 2:
   sharded_tensor.empty(....)
```

Its probably cleaner if all ranks just call the sharded_tensor constructor and ranks which are not part of the process group just create an empty ShardedTensor. `_barrier` doesn't support a mode where `rpc_workers` is not specified, that is why I had to use `_all_gather` instead.

wanchaol commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -48,3 +49,39 @@ def empty(
         memory_format=memory_format,
         process_group=process_group)
+
+def state_dict_hook(module, destination, prefix, local_metadata):
```
why we need to have a static_dict_hook to update the state_dict? could we directly adding the sharded tensor to `nn.module.state_dict`? I feel like it's weird that user need to register a hook in order to save a sharded tensor, ideally user would not notice it as sharded tensor is just another form of tensor.

pritamdamania87 replied:
I couldn't find a better way to add ShardedTensor to the state_dict. ShardedTensor is not a subclass of Tensor so something like `torch.nn.Parameter(my_sharded_tensor)` doesn't work. Also, if you look at `torch.nn.Parameter` it calls `_make_subclass` which detaches the tensor and does a bunch of Tensor specific calls under the hood. Even if we make ShardedTensor to be a subclass of Tensor it is not clear what this Tensor is since ShardedTensor can actually have a list of local shards.

The PyTorch core team is aware of the problem and they are planning to design a brand new API to support this in the future: https://github.com/pytorch/pytorch/issues/62094. But for now, it looks like this hook approach is the best way forward.

wanchaol reviewed (commented):
Have some concerns on using the hook to update the state_dict, otherwise looks good to me!

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -432,3 +472,71 @@ def remote_shards(self) -> Dict[int, List[rpc.RRef[Shard]]]:
+        # Validate process group.
+        local_rank = distributed_c10d.get_rank(self._process_group)
+        if pg_state.local_rank != local_rank:
+            raise RuntimeError(
```
Can we also add tests to test these errors are correctly throwing?

wanchaol reviewed (approved):
looks great! just one suggestion about adding the tests for exception handling.

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support saving and loading ShardedTensor."


1) Add a state_dict hook to ensure ShardedTensors are
added to a state_dict.
2) Add a pre load state_dict hook to ensure ShardedTensor are added back to a
module at load time.
3) Add a `with_load_process_group` context manager for load time.
4) Added ser-de capability to ShardedTensor.

Differential Revision: [D29927881](https://our.internmc.facebook.com/intern/diff/D29927881/)

[ghstack-poisoned]

another-pjohnson commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -59,6 +78,42 @@ class ShardedTensorMetadata(object):
+        elif self.memory_format == torch.channels_last:
+            mem_format_encoding = 1
+        elif self.memory_format == torch.preserve_format:
+            mem_format_encoding = 1
```
I think I've found a copy-paste error, it seems like from the __setstate__ logic,  mem_format_encoding should be = 2 in this case.

awaelchli commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -78,3 +82,39 @@ def init_from_local_shards(
+    """
+    _recurse_update_module(module, state_dict, prefix)
+
+def _recurse_update_module(module, state_dict, prefix):
+    for attr_name, attr in module.__dict__.items():
+        key = prefix + attr_name
+        if key in state_dict:
+            if isinstance(state_dict[key], ShardedTensor):
+                setattr(module, attr_name, state_dict[key])
+
+    for submodule_name, submodule in module.named_modules():
+        key = prefix + submodule_name
+        if submodule_name:
+            _recurse_update_module(submodule, state_dict, key + '.')
+
+
+def _recurse_update_dict(module, destination, prefix):
+    for attr_name, attr in module.__dict__.items():
+        if isinstance(attr, ShardedTensor):
+            destination[prefix + attr_name] = attr
+
+    for submodule_name, submodule in module.named_modules():
+        if submodule_name != '':
+            _recurse_update_dict(submodule, destination, prefix + submodule_name + '.')
```
I posted an issue here #68805 for a potential improvement. I believe the recursion here is not necessary and causes inefficiency when retrieving the state dict.
